src/audio/kpb.c

```diff
@@ -1716,6 +1779,36 @@ static inline bool kpb_is_sample_width_supported(uint32_t sampling_width)
 	return ret;
 }
 
+void kpb_copy_24b_in_32b(const struct audio_stream __sparse_cache *source, uint32_t ioffset,
+			 struct audio_stream __sparse_cache *sink, uint32_t ooffset,
+			 uint32_t n_samples)
+{
+	int ssize = audio_stream_sample_bytes(source); /* src fmt == sink fmt */
+	uint8_t *in = audio_stream_wrap(source, (uint8_t *)source->r_ptr + ioffset * ssize);
+	uint8_t *out = audio_stream_wrap(sink, (uint8_t *)sink->w_ptr + ooffset * ssize);
+
+	const ae_int32x2 *sin = (const ae_int32x2 *)(in);
+	ae_int32x2 *sout = (ae_int32x2 *)(out);
+	ae_int32x2 vs = AE_ZERO32();
+
+	if (!IS_ALIGNED((uintptr_t)sin, 8)) {
+		AE_L32_IP(vs, (const ae_int32 *)(sin), 4);
+		AE_S32_L_IP(vs, (ae_int32 *)(sout), 4);
+		n_samples--;
+	}
+	ae_valign align_out = AE_ZALIGN64();
+
+	for (size_t i = 0; i < n_samples / 2; i++) {
```

Comment:
re avoid divisions, this can be shifted which is cheap on xtensa.

Reply:
need to update, this variable is also not changed in the loops so could be calculated before the loop (again compiler should know better, but is not great so lets help it).